processor/src/chiplets/bus/mod.rs

```diff
@@ -58,25 +59,61 @@ impl ChipletsBus {
+    /// executors requesting one or more hash operations for the Stack where all operation lookups
+    /// must be included at the same cycle. For simple permutations this will require 2 lookups,
+    /// while for a Merkle root update it will require 4, since two Hash operations are required.
+    pub fn request_hasher_operation(&mut self, lookups: &[HasherLookup], cycle: usize) {
+        debug_assert!(
+            lookups.len() == 2 || lookups.len() == 4,
+            "incorrect number of lookup rows for hasher operation request"
+        );
+        self.request_lookup(cycle);
+        self.request_rows
+            .push(ChipletsLookupRow::HasherMulti(lookups.to_vec()));
     }
 
-    /// Provides the data of the specified memory access.  When `old_word` and `new_word` are the
-    /// same in the MemoryLookup, this is a read request. When they are different, it's a write  
-    /// request. The memory access data is provided at cycle `response_cycle`, which is the row of
-    /// the execution trace that contains this Memory row.
-    pub fn provide_memory_operation(&mut self, lookup: MemoryLookup, response_cycle: usize) {
-        self.provide_operation(response_cycle);
-        self.response_rows.push(ChipletsLookupRow::Memory(lookup));
+    /// Requests the specified lookup from the Hash Chiplet at the specified `cycle`. Single lookup
+    /// requests are expected to originate from the decoder during control block decoding. This
+    /// lookup can be for either the initial or the final row of the hash operation.
+    pub fn request_hasher_lookup(&mut self, lookup: HasherLookup, cycle: usize) {
+        self.request_lookup(cycle);
+        self.request_rows.push(ChipletsLookupRow::Hasher(lookup));
+    }
+
+    /// Adds the request for the specified lookup to a queue from which it can be sent later when
+    /// the cycle of the request is known. Queued requests are expected to originate from the
+    /// decoder, since the hash is computed at the start of each control block (along with all
+    /// required lookups), but the decoder does not request intermediate and final lookups until the
+    /// end of the control block or until a `RESPAN`, in the case of `SPAN` blocks with more than
+    /// one operation batch.
+    pub fn enqueue_hasher_request(&mut self, lookup: HasherLookup) {
+        self.queued_requests.push(lookup);
+    }
+
+    /// Pops the top HasherLookup request off the queue and sends it to the bus. This request is
+    /// expected to originate from the decoder as it continues or finalizes control blocks with
+    /// `RESPAN` or `END`.
+    pub fn send_queued_hasher_request(&mut self, cycle: usize) {
+        let lookup = self.queued_requests.pop();
+        debug_assert!(lookup.is_some(), "no queued requests");
+
+        if let Some(lookup) = lookup {
+            self.request_hasher_lookup(lookup, cycle);
+        }
+    }
```

Comment:
This is probably a comment for the future: it may be possible to avoid managing the queue of requests explicitly like this (and get rid of `self.queued_requests`).

For example, instead of building both lookups as a part of `merge_control_blocks`, the first lookup would be still built there, but the second lookup would be built (or retrieved from the hasher) as a part of `request_control_block_hash`. This would be much easier if the lookups were "light" (i.e., didn't carry hasher state info) - but even with the current structure, it should be possible.

Reply:
I agree with this, but I would leave it for the future (as noted). The main reason I didn't go down this path was to keep things as simple as possible without too many changes for this first version, so I didn't want to refactor the decoder and hasher too much. I'll leave a TODO inline for this as well.